### PR TITLE
[Commercial Server] InferenceTiming + TimingBuilder

### PR DESCRIPTION
**Pull request by dbogunowicz**
# UI

`TimingBuilder` is an object that aggregates the durations (time deltas in seconds) of various components of the inference pipeline. Once all the desired time deltas are aggregated, the object may build the data structure (dict: [str, float]) that summarizes the collected information.
Example flow:
  ```python
  builder = TimingBuilder()
  builder.start("total_inference")
  builder.start("pre_process")
  do_pre_process()
  builder.stop("pre_process")
  builder.start("engine_forward")
  do_engine_forward()
  builder.stop("engine_forward")
  builder.start("post_process")
  do_post_process()
  builder.stop("post_process")
  builder.stop("total_inference")
  ...
  summary = builder.build()
  print(summary)
  >> {'pre_process: 0.2, `post-process`: 0.2, 'engine_forward': 0.6, 'total_inference': 1}
  ```
 # Testing
 PR is covered by its respective unittest

### CODE REVIEW

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:149]:
```suggestion
    def build(self) -> InferenceTimingSchema:
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:136]:
```suggestion
    def post_process_delta(self, value: float):
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:127]:
```suggestion
    def engine_forward_delta(self, value: float):
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:118]:
```suggestion
    def pre_process_delta(self, value: float):
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:101]:
```suggestion
    def t1(self, value : float):
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:84]:
```suggestion
    def t0(self, value: float):
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:80]:
```suggestion
    def post_process_delta(self) -> float:
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:76]:
```suggestion
    def engine_forward_delta(self) -> float:
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:72]:
```suggestion
    def pre_process_delta(self) -> float:
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:68]:
```suggestion
    def t1(self) -> float:
```

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:64]:
```suggestion
    def t0(self) -> float:
```

File: src/deepsparse/timing/timing_builder.py
Comment (bfineran) [inline src/deepsparse/timing/timing_builder.py:51]: thoughts on the following?

1. Removing the specific `_start()` and `_complete()` methods in favor of `builder.start_phase(name)`, `builder.end_phase(name)`.  
2. Keeping `start_times` and `end_times` dicts internally that are populated by `start/end_phase` functions
3. measuring deltas between the dict keys on build, and passing in as kwargs to the timing object

don't think that users need access to the deltas before build and this makes it a bit more flexible.  we can either fail if invalid phase names are added, or pipe them in as extras to the timing object

File: src/deepsparse/timing/timing_builder.py
Comment (bfineran) [inline src/deepsparse/timing/timing_builder.py:60]: why do we need this? looks like it adds the potential for annoying bugs if the builder methods are called out of order for some reason.  see longer comment above about refactoring, shouldn't have a need for this after.

File: src/deepsparse/timing/timing_schema.py
Comment (bfineran) [inline src/deepsparse/timing/timing_schema.py:31]: nit: add as pydantic fields w/ description per org standards

**Review by bfineran — CHANGES_REQUESTED** [review]

File: src/deepsparse/timing/timing_builder.py
Comment (KSGulin) [inline src/deepsparse/timing/timing_builder.py:106]: Is there a reason for requiring initialize? Are there plans to expand the logic for initialize in the future?

File: src/deepsparse/timing/timing_builder.py
Comment (KSGulin) [inline src/deepsparse/timing/timing_builder.py:98]:
```suggestion
                f"but its start time is missing"
```

File: src/deepsparse/timing/timing_builder.py
Comment (KSGulin) [inline src/deepsparse/timing/timing_builder.py:86]: I'd recommend using `perf_counter()` over `time()` here, as it provides higher precision. Going back to our HPC class notes, the IEEE standard for floating points allocates a lot more floating point representation density to values near 1, so the large values you get with `time()` fall in a low resolution range. `perf_counter()` gets around that having a much later epoch (and thus returning smaller values)

Comment (corey-nm) [reply src/deepsparse/timing/timing_builder.py:86]: +1

File: src/deepsparse/timing/timing_builder.py
Comment (corey-nm) [inline src/deepsparse/timing/timing_builder.py:67]: Can we get this split into `self.start_times = {}` and `self.stop_times = {}`? I think it'd make it slightly easier to validate that something has started, and slightly easier to read

File: src/deepsparse/timing/timing_builder.py
Comment (corey-nm) [inline src/deepsparse/timing/timing_builder.py:114]: What's the reasoning for using pydantic here? This is all an internal api, so assuming its tested, it should never create fields that shouldn't be there. 

An alternative to this approach is to have a `def query(self, phase) -> time` method. Then you wouldn't have to return anything and wouldn't need this extra step.

File: tests/deepsparse/timing/test_timing_builder.py
Comment (corey-nm) [inline tests/deepsparse/timing/test_timing_builder.py:26]: ❤️ 😁

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [reply src/deepsparse/timing/timing_builder.py:106]: Good point, I do not really see a reason why it is required, will remove it.

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [reply src/deepsparse/timing/timing_builder.py:114]: Not entirely sure what you mean here. My goal was to use TimerBuilder to be able to:
-  create time deltas for phases with different names
-  once the deltas area is created, put them into a schema to be returned for further consumption

I feel that using Pydantic schema here prevents us from e.g. having a typo in one of our `phase_name`s.
E.g. `pre_processing` could be erronously written as `pre_processing_time`.  Pydantic would catch that.

Comment (corey-nm) [reply src/deepsparse/timing/timing_builder.py:114]: Ahh composition gotcha, that makes sense. Right cause these will be summed/averaged together from the different runs? Okay yeah this makes more sense then!

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [inline src/deepsparse/timing/timing_builder.py:39]:
```suggestion
```

File: src/deepsparse/timing/timing_builder.py
Comment (corey-nm) [inline src/deepsparse/timing/timing_builder.py:99]: Can we just return InferenceTimingSchema here instead of passing in a schema?

Comment (corey-nm) [reply src/deepsparse/timing/timing_builder.py:99]: Alternatively going back to dictionary comment - this class could just return the delta dictionary, and then whatever class is using the TimingBuilder could build it's schema using the dictionary. That way this TimingBuilder class doesn't have to know anything about schemas, but we are still capturing that type checking outside of this class

**Review by KSGulin — DISMISSED** [review]
LGTM pending @corey-nm's comment

File: src/deepsparse/timing/timing_builder.py
Comment (dbogunowicz) [reply src/deepsparse/timing/timing_builder.py:99]: Why don't you like the current state of the `build` method?

I could agree on returning the `InferenceTimingSchema` and not passing a `schema` as an arg.
We will be effectively limiting the flexibility of the object (you cannot name your phases however you like), but maybe for a good cause.
And then if we want to extend the usability of the TimingBuilder in the future, we can do it (and then perhaps return dicts as you proposed).

Comment (corey-nm) [reply src/deepsparse/timing/timing_builder.py:99]: I think because this method is doing some validating on fields you used earlier, whereas the start/stop methods don't do that validation. So there is a little bit of inconsistency.

The phase names and therefore fields of the schema are controlled outside of this class, so in my head it makes sense for building the schema to be outside of this class as well.

Agree that returning InferenceTimingSchema from this function does seem too limiting with that in mind (since the phase names are not required to be valid for InferenceTimingSchema).

Comment (dbogunowicz) [reply src/deepsparse/timing/timing_builder.py:99]: Ok, now I am leaning towards returning the dict and putting it into the schema outside of the class methods.

Comment (bfineran) [reply src/deepsparse/timing/timing_builder.py:99]: agreed, I like this flow - it doesn't intrude on anything we'll need in the server and keeps the builder as something we can use generically for other timings

**Review by bfineran — DISMISSED** [review]

**Review by bfineran — APPROVED** [review]
LGTM, please update PR description with goal, ux, test plan before landing

**Review by KSGulin — APPROVED** [review]

**Pull request merged**
